src/db.rs

```diff
@@ -38,315 +39,338 @@ impl Default for Connection {
-    let policy_id: Option<u64> = trnsx.exec_first(mqstr, (keysetid,))?;
+    let polid_row = sqlx::query(query.as_str()).fetch_one(&dbpool).await?;
 
-    if let Some(id) = policy_id {
-        Ok(Some(get_policy(id)?))
-    } else {
-        Ok(None)
-    }
+    let keyset_policy_id = polid_row.try_get::<i32, _>(0)? as u64;
+
+    let retpol = get_policy(keyset_policy_id).await?;
+
+    Ok(retpol)
 }
 
-// A keyset is just a group of keys. This function returns a vector of key ids for secrets in table secrets
-pub fn get_keyset_ids(keysetid: &str) -> Result<Vec<String>> {
-    let mut dbconn = get_dbconn()?;
-    let mut trnsx = dbconn.start_transaction(TxOpts::default())?;
+pub async fn get_keyset_ids(keyset_id: &str) -> Result<Vec<String>> {
+    let dbpool = get_dbpool().await?;
 
-    let mqstr = "SELECT kskeys FROM keysets WHERE keysetid = ?";
-    let keyset_json_opt: Option<String> = trnsx.exec_first(mqstr, (keysetid,))?;
-    let keyset_json_str =
-        keyset_json_opt.ok_or_else(|| anyhow!("db::get_keyset_ids- keyset id not found"))?;
+    let query = format!(
+        "SELECT kskeys FROM keysets WHERE keysetid = \'{}\'",
+        keyset_id
+    );
 
-    let rks: Vec<String> = serde_json::from_str(&keyset_json_str).unwrap();
+    let keyset_row = sqlx::query(query.as_str()).fetch_one(&dbpool).await?;
+
+    let rks: Vec<String> = serde_json::from_str(&keyset_row.try_get::<String, _>(0)?).unwrap();
 
     Ok(rks)
 }
 
-pub fn get_secret(secret_id: &str) -> Result<request::Key> {
-    let mut dbconn = get_dbconn()?;
-    let mut trnsx = dbconn.start_transaction(TxOpts::default())?;
+pub async fn get_secret(secret_id: &str) -> Result<request::Key> {
+    let dbpool = get_dbpool().await?;
+
+    let query = format!(
+        "SELECT secret FROM secrets WHERE secret_id = \'{}\'",
+        secret_id
+    );
 
-    let mqstr = "SELECT secret FROM secrets WHERE secret_id = ?";
-    let val: Option<String> = trnsx.exec_first(mqstr, (secret_id,))?;
+    let payload_row = sqlx::query(query.as_str()).fetch_one(&dbpool).await?;
 
-    let payload = val.ok_or_else(|| anyhow!("secret not found."))?;
+    let payload = payload_row.try_get::<String, _>(0)?;
 
     Ok(request::Key {
         id: secret_id.to_string(),
         payload,
     })
 }
 
-pub fn insert_secret(secret_id: &str, secret: &str, policy_id: Option<u64>) -> Result<()> {
-    let mut dbconn = get_dbconn()?;
+pub async fn insert_secret(secret_id: &str, secret: &str, policy_id: Option<u64>) -> Result<()> {
+    let dbpool = get_dbpool().await?;
 
-    let mqstr = "INSERT INTO secrets (secret_id, secret, polid ) VALUES(?, ?, ?)";
+    let query = match policy_id {
+        Some(p) => format!(
+            "INSERT INTO secrets (secret_id, secret, polid ) VALUES(\'{}\', \'{}\', {:?})",
+            secret_id, secret, p
+        ),
+        None => format!(
+            "INSERT INTO secrets (secret_id, secret ) VALUES(\'{}\', \'{}\')",
+            secret_id, secret
+        ),
+    };
```

Comment:
This was quiet elegant before and now, well, not so much. I wonder if there's an sqlx trick to make this behave like the mysql crate (if the option is None, use a db NULL value).

I warned against `format!`, but one possible solution (if we must use `format!`) is something like (untested):

```rust
let query = format!(
            "INSERT INTO secrets (secret_id, secret, polid ) VALUES(\'{}\', \'{}\', {:?})",
            secret_id, secret, p.unwrap_or("NULL")
        );
```

Reply:
I understand that this was quit elegant before and wish that I could have found an equally elegant solution with format, but this one won't work.  Once we specify polid as a field int the query, sqlx will look for a u64 and not a string when policy_id has a value of None.  As written, this will not compile.  If we want to designate a policy value of 0 as equivalent to NULL, then we could do this with a more complicated query or simply change the policy_id.unwrap_or("NULL") to policy_id.unwrap_or(0u64).